src/execution/mod.rs

```diff
@@ -104,48 +104,58 @@ impl CallInfo {
         )
     }
 
-    /// Yields the contract calls in DFS (preorder).
+    /// Returns the contract calls in DFS (preorder).
     pub fn gen_call_topology(&self) -> Vec<CallInfo> {
         let mut calls = Vec::new();
+        calls.push(self.clone());
         if self.internal_calls.is_empty() {
-            calls.push(self.clone())
+            return calls;
```

Comment:
this branch seems redundant, maybe just put the else here and change the if to !is_empty

i also think the clone is not needed at all:

```rust
        if !self.internal_calls.is_empty() {
            for call_info in &self.internal_calls {
                calls.extend(call_info.gen_call_topology());
            }
        }
```

Reply:
Both good catches. It looks cleaner that way :)

Reply:
Even better:
```rust
        for call_info in &self.internal_calls {
            calls.extend(call_info.gen_call_topology());
        }
```
Iterating through the empty vector doesn't iterate at all, so the if can be removed.

Reply:
Thanks